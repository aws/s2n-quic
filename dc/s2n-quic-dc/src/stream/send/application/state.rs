// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{
    credentials::Credentials,
    crypto::seal,
    packet::stream::{self, encoder},
    stream::{
        packet_number,
        send::{application::transmission, error::Error, flow, path, probes},
        TransportFeatures,
    },
};
use bytes::buf::UninitSlice;
use s2n_codec::EncoderBuffer;
use s2n_quic_core::{
    buffer::{self, reader::Storage as _, Reader as _},
    ensure,
    time::Clock,
    varint::VarInt,
};

pub trait Message {
    fn max_segments(&self) -> usize;
    fn push<P: FnOnce(&mut UninitSlice) -> transmission::Event<()>>(
        &mut self,
        buffer_len: usize,
        p: P,
    );
}

#[derive(Clone, Copy, Debug)]
pub struct State {
    pub is_reliable: bool,
}

impl State {
    #[inline]
    pub fn transmit<E, I, Clk, M>(
        &self,
        credits: flow::Credits,
        path: &path::Info,
        storage: &mut I,
        packet_number: &packet_number::Counter,
        encrypt_key: &E,
        credentials: &Credentials,
        stream_id: &stream::Id,
        source_queue_id: Option<VarInt>,
        clock: &Clk,
        message: &mut M,
        features: &TransportFeatures,
    ) -> Result<(), Error>
    where
        E: seal::Application,
        I: buffer::reader::Storage<Error = core::convert::Infallible>,
        Clk: Clock,
        M: Message,
    {
        ensure!(
            credits.len > 0 || storage.buffer_is_empty() || credits.is_fin,
            Ok(())
        );

        let mut reader = buffer::reader::Incremental::new(credits.offset);
        let mut reader = reader.with_storage(storage, credits.is_fin)?;
        debug_assert!(
            reader.buffered_len() >= credits.len,
            "attempted to acquire more credits than what is buffered"
        );
        let mut reader = reader.with_read_limit(credits.len);

        let max_header_len = self.max_header_len();

        let mut total_payload_len = 0;
        let max_record_size = if features.is_stream() {
            // If the underlying transport is stream based, it will perform its own packetization
            // based on the MTU determined at that layer. Therefore, we do not need to restrict
            // writes to the probed max datagram size, and can instead use a larger value, in this
            // case 2^14, based on the TLS max record size.
            1 << 14
        } else {
            path.max_datagram_size
        };

        loop {
            let packet_number = packet_number.next()?;

            let buffer_len = {
                let estimated_len = reader.buffered_len() + max_header_len;
                (max_record_size as usize).min(estimated_len)
            };

            message.push(buffer_len, |buffer| {
                let stream_offset = reader.current_offset();
                let mut reader = reader.track_read();

                let buffer = unsafe {
                    // SAFETY: `buffer` is a valid `UninitSlice` but `EncoderBuffer` expects to
                    // write into a `&mut [u8]`. Here we construct a `&mut [u8]` since
                    // `EncoderBuffer` never actually reads from the slice and only writes to it.
                    core::slice::from_raw_parts_mut(buffer.as_mut_ptr(), buffer.len())
                };
                let encoder = EncoderBuffer::new(buffer);
                let packet_len = encoder::encode(
                    encoder,
                    source_queue_id,
                    *stream_id,
                    packet_number,
                    path.next_expected_control_packet,
                    VarInt::ZERO,
                    &mut &[][..],
                    VarInt::ZERO,
                    &(),
                    &mut reader,
                    encrypt_key,
                    credentials,
                );

                // buffer is clamped to u16::MAX so this is safe to cast without loss
                let _: u16 = max_record_size;
                let packet_len = packet_len as u16;
                let payload_len = reader.consumed_len() as u16;
                total_payload_len += payload_len as usize;

                let has_more_app_data = credits.initial_len > total_payload_len;

                let included_fin = reader
                    .final_offset()
                    .is_some_and(|fin| stream_offset.as_u64() + payload_len as u64 == fin.as_u64());

                let time_sent = clock.get_time();
                probes::on_transmit_stream(
                    credentials.id,
                    credentials.key_id,
                    stream::PacketSpace::Stream,
                    s2n_quic_core::packet::number::PacketNumberSpace::Initial
                        .new_packet_number(packet_number),
                    stream_offset,
                    payload_len,
                    included_fin,
                    false,
                );

                let info = transmission::Info {
                    packet_len,
                    retransmission: if stream_id.is_reliable {
                        Some(())
                    } else {
                        None
                    },
                    stream_offset,
                    payload_len,
                    included_fin,
                    time_sent,
                    ecn: path.ecn,
                };

                transmission::Event {
                    packet_number,
                    info,
                    has_more_app_data,
                }
            });

            // bail if we've transmitted everything
            ensure!(!reader.buffer_is_empty(), break);
        }

        Ok(())
    }

    #[inline]
    pub fn max_header_len(&self) -> usize {
        if self.is_reliable {
            encoder::MAX_RETRANSMISSION_HEADER_LEN
        } else {
            encoder::MAX_HEADER_LEN
        }
    }
}
