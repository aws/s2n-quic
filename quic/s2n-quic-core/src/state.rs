// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use core::fmt;

pub type Result<T> = core::result::Result<(), Error<T>>;

#[cfg(feature = "state-tracing")]
#[doc(hidden)]
pub use tracing::debug as _debug;

#[cfg(not(feature = "state-tracing"))]
#[doc(hidden)]
pub use crate::__tracing_noop__ as _debug;

#[macro_export]
#[doc(hidden)]
macro_rules! __state_transition__ {
    ($state:ident, $valid:pat => $target:expr) => {
        $crate::state::transition!(_, $state, $valid => $target)
    };
    ($event:ident, $state:ident, $valid:pat => $target:expr) => {{
        if *$state == $target {
            // if transitioning to itself is not called out, then return an error
            $crate::ensure!(
                matches!($state, $valid),
                Err($crate::state::Error::NoOp { current: $target })
            );
            return Ok(());
        }

        // make sure the transition is valid
        $crate::ensure!(
            matches!($state, $valid),
            Err($crate::state::Error::InvalidTransition {
                current: $state.clone(),
                target: $target
            })
        );

        let __event__ = stringify!($event);
        if __event__.is_empty() || __event__ == "_" {
            $crate::state::_debug!(prev = ?$state, next = ?$target);
        } else {
            $crate::state::_debug!(event = %__event__, prev = ?$state, next = ?$target);
        }

        *$state = $target;
        Ok(())
    }};
}

pub use crate::__state_transition__ as transition;

#[macro_export]
#[doc(hidden)]
macro_rules! __state_event__ {
    ($(#[doc = $doc:literal])* $event:ident ($($valid:ident)|* => $target:ident)) => {
        $(
            #[doc = $doc]
        )*
        #[inline]
        pub fn $event(&mut self) -> $crate::state::Result<Self> {
            $crate::state::transition!($event, self, $(Self::$valid)|* => Self::$target)
        }
    };
    ($( $(#[doc = $doc:literal])* $event:ident ($($valid:ident)|* => $target:ident) ;)*) => {
        $(
            $crate::state::event!($(#[doc = $doc])* $event($($valid)|* => $target));
        )*

        /// Generates a dot graph of all state transitions
        pub fn dot() -> impl ::core::fmt::Display {
            struct Dot;

            impl ::core::fmt::Display for Dot {
                fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
                    writeln!(f, "digraph {{")?;

                    let mut all_states = [
                        // collect all of the states we've observed
                        $(
                            $(
                                stringify!($valid),
                            )*
                            stringify!($target),
                        )*
                    ];

                    let all_states = $crate::state::dedup_states(&mut all_states);

                    for state in all_states {
                        writeln!(f, "  {state};")?;
                    }

                    $(
                        $(
                            writeln!(
                                f,
                                "  {} -> {} [label = {:?}];",
                                stringify!($valid),
                                stringify!($target),
                                stringify!($event),
                            )?;
                        )*
                    )*

                    writeln!(f, "}}")?;
                    Ok(())
                }
            }

            Dot
        }
    }
}

pub use crate::__state_event__ as event;

#[doc(hidden)]
pub fn dedup_states<'a>(states: &'a mut [&'a str]) -> &'a mut [&'a str] {
    // TODO replace with
    // https://doc.rust-lang.org/std/primitive.slice.html#method.partition_dedup
    // when stable
    //
    // For now, we've just inlined their implementation

    let len = states.len();
    if len <= 1 {
        return states;
    }

    // sort the states before deduping
    states.sort_unstable();

    let ptr = states.as_mut_ptr();
    let mut next_read: usize = 1;
    let mut next_write: usize = 1;

    // SAFETY: the `while` condition guarantees `next_read` and `next_write`
    // are less than `len`, thus are inside `self`. `prev_ptr_write` points to
    // one element before `ptr_write`, but `next_write` starts at 1, so
    // `prev_ptr_write` is never less than 0 and is inside the slice.
    // This fulfils the requirements for dereferencing `ptr_read`, `prev_ptr_write`
    // and `ptr_write`, and for using `ptr.add(next_read)`, `ptr.add(next_write - 1)`
    // and `prev_ptr_write.offset(1)`.
    //
    // `next_write` is also incremented at most once per loop at most meaning
    // no element is skipped when it may need to be swapped.
    //
    // `ptr_read` and `prev_ptr_write` never point to the same element. This
    // is required for `&mut *ptr_read`, `&mut *prev_ptr_write` to be safe.
    // The explanation is simply that `next_read >= next_write` is always true,
    // thus `next_read > next_write - 1` is too.
    unsafe {
        // Avoid bounds checks by using raw pointers.
        while next_read < len {
            let ptr_read = ptr.add(next_read);
            let prev_ptr_write = ptr.add(next_write - 1);
            if *ptr_read != *prev_ptr_write {
                if next_read != next_write {
                    let ptr_write = prev_ptr_write.add(1);
                    core::ptr::swap(ptr_read, ptr_write);
                }
                next_write += 1;
            }
            next_read += 1;
        }
    }

    &mut states[..next_write]
}

#[macro_export]
#[doc(hidden)]
macro_rules! __state_is__ {
    ($(#[doc = $doc:literal])* $function:ident, $($state:ident)|+) => {
        $(
            #[doc = $doc]
        )*
        #[inline]
        pub fn $function(&self) -> bool {
            matches!(self, $(Self::$state)|*)
        }
    };
}

pub use crate::__state_is__ as is;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Error<T> {
    NoOp { current: T },
    InvalidTransition { current: T, target: T },
}

impl<T: fmt::Debug> fmt::Display for Error<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::NoOp { current } => {
                write!(f, "state is already set to {current:?}")
            }
            Self::InvalidTransition { current, target } => {
                write!(f, "invalid transition from {current:?} to {target:?}",)
            }
        }
    }
}

#[cfg(feature = "std")]
impl<T: fmt::Debug> std::error::Error for Error<T> {}
