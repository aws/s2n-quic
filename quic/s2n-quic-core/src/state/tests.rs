// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::*;
use insta::{assert_debug_snapshot, assert_snapshot};

#[derive(Clone, Debug, Default, PartialEq, Eq)]
enum State {
    #[default]
    Init,
    Left,
    Right,
    LeftLeft,
    LeftRight,
    RightLeft,
    RightRight,
}

impl State {
    event! {
        on_left(
            Init => Left,
            Left => LeftLeft,
            Right => RightLeft,
        );
        on_right(
            Init => Right,
            Left => LeftRight,
            Right => RightRight,
        );
    }
}

#[test]
#[cfg_attr(miri, ignore)]
fn snapshots() {
    let mut outcomes = vec![];
    let states = [
        State::Init,
        State::Left,
        State::Right,
        State::LeftLeft,
        State::LeftRight,
        State::RightLeft,
        State::RightRight,
    ];
    for state in states {
        macro_rules! push {
            ($event:ident) => {
                let mut target = state.clone();
                let result = target.$event().map(|_| target);
                outcomes.push((state.clone(), stringify!($event), result));
            };
        }
        push!(on_left);
        push!(on_right);
    }

    assert_debug_snapshot!(outcomes);
}

#[test]
#[cfg_attr(miri, ignore)]
fn dot_test() {
    assert_snapshot!(State::dot());
}
