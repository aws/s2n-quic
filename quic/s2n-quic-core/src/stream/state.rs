// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

mod recv;
mod send;

pub use recv::Receiver;
pub use send::Sender;
