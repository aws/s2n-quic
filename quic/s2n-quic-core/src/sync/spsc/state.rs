// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::{Cell, ClosedError, Result, Slice};
use alloc::alloc::Layout;
use atomic_waker::AtomicWaker;
use core::{
    marker::PhantomData,
    ops::Deref,
    panic::{RefUnwindSafe, UnwindSafe},
    ptr::NonNull,
    sync::atomic::{AtomicBool, AtomicUsize, Ordering},
};

type Pair<'a, T> = super::Pair<Slice<'a, Cell<T>>>;

#[derive(Clone, Copy, Debug)]
pub struct Cursor {
    head: usize,
    tail: usize,
    capacity: usize,
}

impl Cursor {
    #[inline]
    fn new(capacity: usize) -> Self {
        Self {
            head: 0,
            tail: 0,
            capacity,
        }
    }

    #[inline]
    pub fn is_empty(&self) -> bool {
        self.head == self.tail
    }

    #[inline]
    pub fn is_full(&self) -> bool {
        self.recv_len() == self.capacity
    }

    #[inline]
    pub fn recv_len(&self) -> usize {
        self.distance(self.head, self.tail)
    }

    #[inline]
    pub fn send_capacity(&self) -> usize {
        self.capacity - self.recv_len()
    }

    #[inline]
    pub fn increment_head(&mut self, len: usize) {
        self.head = self.increment(self.head, len);
    }

    #[inline]
    pub fn increment_tail(&mut self, len: usize) {
        self.tail = self.increment(self.tail, len);
    }

    #[inline]
    fn wrapped_head(&self) -> usize {
        self.collapse_position(self.head)
    }

    #[inline]
    fn wrapped_tail(&self) -> usize {
        self.collapse_position(self.tail)
    }

    /// Returns the distance between two positions.
    #[inline]
    fn distance(&self, a: usize, b: usize) -> usize {
        unsafe {
            unsafe_assert!(a == 0 || a < 2 * self.capacity);
            unsafe_assert!(b == 0 || b < 2 * self.capacity);
        }
        if a <= b {
            b - a
        } else {
            2 * self.capacity - a + b
        }
    }

    #[inline]
    fn increment(&self, pos: usize, n: usize) -> usize {
        unsafe {
            unsafe_assert!(pos == 0 || pos < 2 * self.capacity);
            unsafe_assert!(n <= self.capacity);
        }
        let threshold = 2 * self.capacity - n;
        if pos < threshold {
            pos + n
        } else {
            pos - threshold
        }
    }

    #[inline]
    fn collapse_position(&self, pos: usize) -> usize {
        unsafe {
            unsafe_assert!(pos == 0 || pos < 2 * self.capacity);
        }
        if pos < self.capacity {
            pos
        } else {
            pos - self.capacity
        }
    }
}

pub struct Ring;

impl Behavior for Ring {}

pub struct DoubleRing;

impl Behavior for DoubleRing {
    #[inline]
    fn header_capacity(capacity: usize) -> usize {
        capacity * 2
    }

    #[inline]
    fn replicate_tail<T>(prev: &Cursor, state: &State<T, Self>) {
        let (front_filled, _) = state.as_pairs();
        let back = unsafe {
            let capacity = state.cursor.capacity;
            let ptr = state.data_ptr().add(capacity);
            core::slice::from_raw_parts(ptr, capacity)
        };
        let (back_filled, _) = state.data_to_pairs(back);

        let mut skip = prev.recv_len();
        let len = prev.distance(prev.tail, state.cursor.tail);

        if len == state.cursor.capacity {
            skip = 0;
        }

        unsafe {
            front_filled.replicate_to(back_filled, skip, len);
        }
    }
}

pub trait Behavior: Sized {
    #[inline]
    fn header_capacity(capacity: usize) -> usize {
        capacity
    }

    #[inline]
    fn replicate_tail<T>(prev: &Cursor, state: &State<T, Self>) {
        let _ = prev;
        let _ = state;
    }
}

#[derive(Debug)]
pub struct State<T, B> {
    header: NonNull<Header<T>>,
    pub cursor: Cursor,
    behavior: PhantomData<B>,
}

unsafe impl<T: Send, B> Send for State<T, B> {}
unsafe impl<T: Sync, B> Sync for State<T, B> {}
impl<T: RefUnwindSafe, B> UnwindSafe for State<T, B> {}

impl<T, B> Clone for State<T, B> {
    #[inline]
    fn clone(&self) -> Self {
        Self {
            header: self.header,
            cursor: self.cursor,
            behavior: PhantomData,
        }
    }
}

impl<T, B> Deref for State<T, B> {
    type Target = Header<T>;

    #[inline]
    fn deref(&self) -> &Self::Target {
        unsafe { self.header.as_ref() }
    }
}

impl<T, B: Behavior> State<T, B> {
    #[inline]
    pub fn new(capacity: usize) -> Self {
        // double the capacity of the header so we can return contiguous regions
        let header_capacity = B::header_capacity(capacity);
        let header = Header::alloc(header_capacity).expect("could not allocate channel");
        let cursor = Cursor::new(capacity);
        Self {
            header,
            cursor,
            behavior: PhantomData,
        }
    }

    #[inline]
    pub fn acquire_capacity(&mut self) -> Result<bool> {
        if !self.open.load(Ordering::Acquire) {
            return Err(ClosedError);
        }

        if !self.cursor.is_full() {
            return Ok(true);
        }

        // update the cached version
        self.cursor.head = self.head.load(Ordering::Acquire);

        let is_full = self.cursor.is_full();

        Ok(!is_full)
    }

    #[inline]
    pub fn acquire_filled(&mut self) -> Result<bool> {
        if !self.cursor.is_empty() {
            return Ok(true);
        }

        self.cursor.tail = self.tail.load(Ordering::Acquire);

        if !self.cursor.is_empty() {
            return Ok(true);
        }

        if !self.open.load(Ordering::Acquire) {
            return Err(ClosedError);
        }

        Ok(false)
    }

    #[inline]
    pub fn persist_head(&self, prev: Cursor) {
        // nothing changed
        if prev.head == self.cursor.head {
            return;
        }

        self.head.store(self.cursor.head, Ordering::Release);
    }

    #[inline]
    pub fn persist_tail(&self, prev: Cursor) {
        // nothing changed
        if prev.tail == self.cursor.tail {
            return;
        }

        B::replicate_tail(&prev, self);

        self.tail.store(self.cursor.tail, Ordering::Release);
    }

    #[inline]
    fn data(&self) -> &[Cell<T>] {
        unsafe {
            let ptr = self.data_ptr();
            // we don't multiply the capacity here for simplicity
            let capacity = self.cursor.capacity;
            core::slice::from_raw_parts(ptr, capacity)
        }
    }

    #[inline]
    fn data_ptr(&self) -> *const Cell<T> {
        unsafe {
            let capacity = B::header_capacity(self.cursor.capacity);
            let (_, offset) = Header::<T>::layout_unchecked(capacity);

            let ptr = self.header.as_ptr() as *const u8;
            let ptr = ptr.add(offset);
            ptr as *const Cell<T>
        }
    }

    #[inline]
    pub fn try_close(&mut self) -> bool {
        let prev = self.open.swap(false, Ordering::Acquire);
        if !prev {
            unsafe {
                self.drop_contents();
            }
        }
        prev
    }

    #[inline]
    pub fn as_pairs(&self) -> (Pair<T>, Pair<T>) {
        let data = self.data();
        self.data_to_pairs(data)
    }

    #[inline]
    fn data_to_pairs<'a>(&self, data: &'a [Cell<T>]) -> (Pair<'a, T>, Pair<'a, T>) {
        if self.cursor.is_full() {
            let head = self.cursor.wrapped_head();
            let (filled_tail, filled_head) = data.split_at(head);
            let filled = Pair {
                head: Slice(filled_head),
                tail: Slice(filled_tail),
            };
            let unfilled = Pair {
                head: Slice(&[]),
                tail: Slice(&[]),
            };
            return (filled, unfilled);
        }

        let head = self.cursor.wrapped_head();
        let tail = self.cursor.wrapped_tail();

        let is_contiguous = tail >= head;

        if is_contiguous {
            let (data, unfilled_head) = data.split_at(tail);
            let (unfilled_tail, filled_head) = data.split_at(head);

            let filled = Pair {
                head: Slice(filled_head),
                tail: Slice(&[]),
            };
            let unfilled = Pair {
                head: Slice(unfilled_head),
                tail: Slice(unfilled_tail),
            };
            (filled, unfilled)
        } else {
            let (data, filled_head) = data.split_at(head);
            let (filled_tail, unfilled_head) = data.split_at(tail);

            let filled = Pair {
                head: Slice(filled_head),
                tail: Slice(filled_tail),
            };
            let unfilled = Pair {
                head: Slice(unfilled_head),
                tail: Slice(&[]),
            };
            (filled, unfilled)
        }
    }

    #[inline]
    unsafe fn drop_contents(&mut self) {
        // refresh the cursor from the shared state
        self.cursor.head = self.head.load(Ordering::Acquire);
        self.cursor.tail = self.tail.load(Ordering::Acquire);

        // release all of the filled data
        let (filled, _unfilled) = self.as_pairs();
        for cell in filled.iter() {
            drop(cell.take());
        }

        // free the header
        let ptr = self.header.as_ptr() as *mut u8;
        let capacity = B::header_capacity(self.cursor.capacity);
        let (layout, _offset) = Header::<T>::layout_unchecked(capacity);
        alloc::alloc::dealloc(ptr, layout)
    }
}

impl<T> State<T, DoubleRing> {
    #[inline]
    pub fn as_slices(&self) -> (Slice<Cell<T>>, Slice<Cell<T>>) {
        let data = self.data();

        if self.cursor.is_full() {
            let head = self.cursor.wrapped_head();
            let filled = Slice(unsafe { data.get_unchecked(head..) });
            let unfilled = Slice(&[]);
            return (filled, unfilled);
        }

        let head = self.cursor.wrapped_head();
        let tail = self.cursor.wrapped_tail();

        let filled = Slice(unsafe { data.get_unchecked(head..tail) });
        let unfilled = Slice(unsafe { data.get_unchecked(tail..) });
        (filled, unfilled)
    }
}

#[derive(Debug)]
pub struct Header<T> {
    head: AtomicUsize,
    pub receiver: AtomicWaker,
    tail: AtomicUsize,
    pub sender: AtomicWaker,
    open: AtomicBool,
    data: PhantomData<T>,
}

impl<T> Header<T> {
    fn alloc(capacity: usize) -> Option<NonNull<Self>> {
        unsafe {
            let (layout, _offset) = Self::layout(capacity).ok()?;
            let state = alloc::alloc::alloc(layout);
            let state = state as *mut Self;
            let state = NonNull::new(state)?;

            state.as_ptr().write(Self::new());

            Some(state)
        }
    }

    #[inline]
    const fn new() -> Self {
        Self {
            head: AtomicUsize::new(0),
            tail: AtomicUsize::new(0),
            sender: AtomicWaker::new(),
            receiver: AtomicWaker::new(),
            open: AtomicBool::new(true),
            data: PhantomData,
        }
    }

    #[inline]
    fn layout(capacity: usize) -> Result<(Layout, usize), alloc::alloc::LayoutError> {
        let header_layout = Layout::new::<Self>();
        let data_layout = Layout::array::<Cell<T>>(capacity)?;
        let (layout, offset) = header_layout.extend(data_layout)?;
        Ok((layout, offset))
    }

    #[inline]
    unsafe fn layout_unchecked(capacity: usize) -> (Layout, usize) {
        if let Ok(v) = Self::layout(capacity) {
            v
        } else {
            core::hint::unreachable_unchecked()
        }
    }
}
