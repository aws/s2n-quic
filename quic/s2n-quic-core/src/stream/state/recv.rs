// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::state::{event, is};

//= https://www.rfc-editor.org/rfc/rfc9000#section-3.2
//#        o
//#       | Recv STREAM / STREAM_DATA_BLOCKED / RESET_STREAM
//#       | Create Bidirectional Stream (Sending)
//#       | Recv MAX_STREAM_DATA / STOP_SENDING (Bidirectional)
//#       | Create Higher-Numbered Stream
//#       v
//#   +-------+
//#   | Recv  | Recv RESET_STREAM
//#   |       |-----------------------.
//#   +-------+                       |
//#       |                           |
//#       | Recv STREAM + FIN         |
//#       v                           |
//#   +-------+                       |
//#   | Size  | Recv RESET_STREAM     |
//#   | Known |---------------------->|
//#   +-------+                       |
//#       |                           |
//#       | Recv All Data             |
//#       v                           v
//#   +-------+ Recv RESET_STREAM +-------+
//#   | Data  |--- (optional) --->| Reset |
//#   | Recvd |  Recv All Data    | Recvd |
//#   +-------+<-- (optional) ----+-------+
//#       |                           |
//#       | App Read All Data         | App Read Reset
//#       v                           v
//#   +-------+                   +-------+
//#   | Data  |                   | Reset |
//#   | Read  |                   | Read  |
//#   +-------+                   +-------+

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub enum Receiver {
    #[default]
    Recv,
    SizeKnown,
    DataRecvd,
    DataRead,
    ResetRecvd,
    ResetRead,
}

impl Receiver {
    is!(is_receiving, Recv);
    is!(is_size_known, SizeKnown);
    is!(is_data_received, DataRecvd);
    is!(is_data_read, DataRead);
    is!(is_reset_received, ResetRecvd);
    is!(is_reset_read, ResetRead);
    is!(is_terminal, DataRead | ResetRead);

    event! {
        on_receive_fin(Recv => SizeKnown);
        on_receive_all_data(SizeKnown => DataRecvd);
        on_app_read_all_data(DataRecvd => DataRead);

        on_reset(Recv | SizeKnown => ResetRecvd);
        on_app_read_reset(ResetRecvd => ResetRead);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::{assert_debug_snapshot, assert_snapshot};

    #[test]
    #[cfg_attr(miri, ignore)]
    fn snapshots() {
        let mut outcomes = vec![];
        let states = [
            Receiver::Recv,
            Receiver::SizeKnown,
            Receiver::DataRecvd,
            Receiver::DataRead,
            Receiver::ResetRecvd,
            Receiver::ResetRead,
        ];
        for state in states {
            macro_rules! push {
                ($event:ident) => {
                    let mut target = state.clone();
                    let result = target.$event().map(|_| target);
                    outcomes.push((state.clone(), stringify!($event), result));
                };
            }
            push!(on_receive_fin);
            push!(on_receive_all_data);
            push!(on_app_read_all_data);
            push!(on_reset);
            push!(on_app_read_reset);
        }

        assert_debug_snapshot!(outcomes);
    }

    #[test]
    #[cfg_attr(miri, ignore)]
    fn dot_test() {
        assert_snapshot!(Receiver::dot());
    }
}
