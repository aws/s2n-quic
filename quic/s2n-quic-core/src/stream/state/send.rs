// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::state::{event, is};

//= https://www.rfc-editor.org/rfc/rfc9000#section-3.1
//#        o
//#       | Create Stream (Sending)
//#       | Peer Creates Bidirectional Stream
//#       v
//#   +-------+
//#   | Ready | Send RESET_STREAM
//#   |       |-----------------------.
//#   +-------+                       |
//#       |                           |
//#       | Send STREAM /             |
//#       |      STREAM_DATA_BLOCKED  |
//#       v                           |
//#   +-------+                       |
//#   | Send  | Send RESET_STREAM     |
//#   |       |---------------------->|
//#   +-------+                       |
//#       |                           |
//#       | Send STREAM + FIN         |
//#       v                           v
//#   +-------+                   +-------+
//#   | Data  | Send RESET_STREAM | Reset |
//#   | Sent  |------------------>| Sent  |
//#   +-------+                   +-------+
//#       |                           |
//#       | Recv All ACKs             | Recv ACK
//#       v                           v
//#   +-------+                   +-------+
//#   | Data  |                   | Reset |
//#   | Recvd |                   | Recvd |
//#   +-------+                   +-------+

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub enum Sender {
    #[default]
    Ready,
    Send,
    DataSent,
    DataRecvd,
    /// An additional state for implementations to separate queueing a RESET_STREAM from actually
    /// sending it
    ResetQueued,
    ResetSent,
    ResetRecvd,
}

impl Sender {
    is!(is_ready, Ready);
    is!(is_sending, Send);
    is!(is_data_sent, DataSent);
    is!(is_data_received, DataRecvd);
    is!(is_reset_queued, ResetQueued);
    is!(is_reset_sent, ResetSent);
    is!(is_reset_received, ResetRecvd);
    is!(is_terminal, DataRecvd | ResetRecvd);

    event! {
        on_send_stream(Ready => Send);
        // we can jump from Ready to DataSent even though the
        // diagram doesn't explicitly highlight this transition
        on_send_fin(Ready | Send => DataSent);
        on_recv_all_acks(DataSent | ResetQueued => DataRecvd);

        on_queue_reset(Ready | Send | DataSent => ResetQueued);
        on_send_reset(Ready | Send | DataSent | ResetQueued => ResetSent);
        on_recv_reset_ack(ResetSent => ResetRecvd);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::{assert_debug_snapshot, assert_snapshot};

    #[test]
    #[cfg_attr(miri, ignore)]
    fn snapshots() {
        let mut outcomes = vec![];
        let states = [
            Sender::Ready,
            Sender::Send,
            Sender::DataSent,
            Sender::DataRecvd,
            Sender::ResetQueued,
            Sender::ResetSent,
            Sender::ResetRecvd,
        ];
        for state in states {
            macro_rules! push {
                ($event:ident) => {
                    let mut target = state.clone();
                    let result = target.$event().map(|_| target);
                    outcomes.push((state.clone(), stringify!($event), result));
                };
            }
            push!(on_send_stream);
            push!(on_send_fin);
            push!(on_queue_reset);
            push!(on_send_reset);
            push!(on_recv_all_acks);
            push!(on_recv_reset_ack);
        }

        assert_debug_snapshot!(outcomes);
    }

    #[test]
    #[cfg_attr(miri, ignore)]
    fn dot_test() {
        assert_snapshot!(Sender::dot());
    }
}
