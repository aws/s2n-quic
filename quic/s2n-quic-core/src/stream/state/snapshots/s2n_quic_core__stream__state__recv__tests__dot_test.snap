---
source: quic/s2n-quic-core/src/stream/state/recv.rs
expression: "Receiver::dot()"
---
digraph {
  DataRead;
  DataRecvd;
  Recv;
  ResetRead;
  ResetRecvd;
  SizeKnown;
  Recv -> SizeKnown [label = "on_receive_fin"];
  SizeKnown -> DataRecvd [label = "on_receive_all_data"];
  DataRecvd -> DataRead [label = "on_app_read_all_data"];
  Recv -> ResetRecvd [label = "on_reset"];
  SizeKnown -> ResetRecvd [label = "on_reset"];
  ResetRecvd -> ResetRead [label = "on_app_read_reset"];
}
